server/src/test/java/org/apache/druid/client/selector/TierSelectorStrategyTest.java

```diff
@@ -260,17 +261,223 @@ public void testServerSelectorStrategyDefaults()
+        client
+    );
+    QueryableDruidServer preferredTierHighPriority = new QueryableDruidServer(
+        new DruidServer("test2", "localhost", null, 0, ServerType.HISTORICAL, "preferred", 1),
+        client
+    );
+    QueryableDruidServer nonPreferredTierLowestPriority = new QueryableDruidServer(
+        new DruidServer("test3", "localhost", null, 0, ServerType.HISTORICAL, "non-preferred", -1),
+        client
+    );
+
+    testPreferredTierSelectorStrategy(
+        new PreferredTierSelectorStrategy(
+            new ConnectionCountServerSelectorStrategy(),
+            new PreferredTierSelectorStrategyConfig("preferred", "lowest")
+        ),
+        preferredTierLowPriority, preferredTierHighPriority, nonPreferredTierLowestPriority
+    );
+  }
+
+  @Test
+  public void testPreferredTierSelectorStrategyWithFallback()
+  {
+    DirectDruidClient client = EasyMock.createMock(DirectDruidClient.class);
+    // Create only non-preferred tier servers with different priorities
+    QueryableDruidServer nonPreferredTierLowPriority = new QueryableDruidServer(
+        new DruidServer("test1", "localhost", null, 0, ServerType.HISTORICAL, "non-preferred", 0),
+        client
+    );
+    QueryableDruidServer nonPreferredTierMediumPriority = new QueryableDruidServer(
+        new DruidServer("test2", "localhost", null, 0, ServerType.HISTORICAL, "non-preferred", 1),
+        client
+    );
+    QueryableDruidServer nonPreferredTierHighPriority = new QueryableDruidServer(
+        new DruidServer("test3", "localhost", null, 0, ServerType.HISTORICAL, "non-preferred", 2),
+        client
+    );
+
+    // Since no preferred tier servers are available, it should fall back to other servers
+    // based on highest priority
+    testPreferredTierSelectorStrategy(
+        new PreferredTierSelectorStrategy(
+            new ConnectionCountServerSelectorStrategy(),
+            new PreferredTierSelectorStrategyConfig("preferred", "highest")
+        ),
+        nonPreferredTierHighPriority, nonPreferredTierMediumPriority, nonPreferredTierLowPriority
+    );
+  }
+
+  @Test
+  public void testPreferredTierSelectorStrategyMixedServers()
+  {
+    DirectDruidClient client = EasyMock.createMock(DirectDruidClient.class);
+    QueryableDruidServer preferredTierLowPriority = new QueryableDruidServer(
+        new DruidServer("test1", "localhost", null, 0, ServerType.HISTORICAL, "preferred", 0),
+        client
+    );
+    QueryableDruidServer preferredTierHighPriority = new QueryableDruidServer(
+        new DruidServer("test2", "localhost", null, 0, ServerType.HISTORICAL, "preferred", 1),
+        client
+    );
+    QueryableDruidServer anotherTierHighPriority = new QueryableDruidServer(
+        new DruidServer("test3", "localhost", null, 0, ServerType.HISTORICAL, "tier1", 2),
+        client
+    );
+    QueryableDruidServer yetAnotherTierMediumPriority = new QueryableDruidServer(
```

Comment:
does this introduce randomness to the tests adding a second server of equal priority? That is at least what it seemed like on my local. I got intermittent test failures for this test because the two servers with priority of 1 were flip flopping in actual return order.

Reply:
it looks like two servers with same priority bring different results with multiple runs. 
fixed by setting different priorities for this cases to make sure the test result is stable.